File: app/src/main/java/ru/ra66it/updaterforspotify/notification/PollService.java
package ru.ra66it.updaterforspotify.notification;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.job.JobInfo;
import android.app.job.JobParameters;
import android.app.job.JobScheduler;
import android.app.job.JobService;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Build;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import ru.ra66it.updaterforspotify.MyApplication;
import ru.ra66it.updaterforspotify.R;
import ru.ra66it.updaterforspotify.model.FullSpotifyModel;
import ru.ra66it.updaterforspotify.model.Spotify;
import ru.ra66it.updaterforspotify.rest.SpotifyApi;
import ru.ra66it.updaterforspotify.storage.QueryPreferneces;
import ru.ra66it.updaterforspotify.ui.activity.MainActivity;
import ru.ra66it.updaterforspotify.utils.UtilsSpotify;

/**
 * Created by 2Rabbit on 07.01.2018.
 */


public class PollService extends JobService {

    private static final String TAG = PollService.class.getSimpleName();
    private static final long POLL_INTERVAL = TimeUnit.DAYS.toMillis(1);
    private static final int JOB_ID = 123;
    public static final String NOTIFICATION_ID = "notification_id";
    public static final String NOTIFICATION_CHANEL = "UFS_CHANEL_ID";
    public static final String LATEST_LINK = "latest_link";
    public static final String LATEST_VERSION = "latest_version";
    public static final String LATEST_VERSION_NAME = "latest_version_name";

    private CompositeDisposable compositeDisposable;
    private FullSpotifyModel fullSpotifyModel;

    @Inject
    SpotifyApi spotifyApi;

    @Override
    public void onCreate() {
        super.onCreate();
        MyApplication.getApplicationComponent().inject(this);
        Log.i(TAG, "Received an JobService");
        compositeDisposable = new CompositeDisposable();
    }

    public static void setServiceAlarm(Context context, boolean isOn) {
        ComponentName component = new ComponentName(context, PollService.class);
        JobInfo builder;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            builder = new JobInfo.Builder(JOB_ID, component)
                    .setMinimumLatency(POLL_INTERVAL)
                    .setOverrideDeadline(POLL_INTERVAL)
                    .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
                    .setPersisted(true)
                    .build();
        } else {
            builder = new JobInfo.Builder(JOB_ID, component)
                    .setPeriodic(POLL_INTERVAL)
                    .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
                    .setPersisted(true)
                    .build();
        }

        JobScheduler jobScheduler = (JobScheduler) context.getSystemService(Context.JOB_SCHEDULER_SERVICE);

        if (isOn) {
            jobScheduler.cancel(JOB_ID);
            jobScheduler.schedule(builder);
        } else {
            jobScheduler.cancelAll();
        }

    }

    @Override
    public boolean onStartJob(JobParameters jobParameters) {
        if (QueryPreferneces.getNotificationDogFood(this)) {
            notificationsSpotifyDF(jobParameters);
        } else if (QueryPreferneces.getNotificationDogFoodC(this)) {
            notificationsSpotifyDFC(jobParameters);
        } else if (QueryPreferneces.getNotificationOrigin(this)) {
            if (QueryPreferneces.isSpotifyBeta(this)) {
                notificationsSpotifyOrigBeta(jobParameters);
            } else {
                notificationsSpotifyOrig(jobParameters);
            }
        }

        return true;
    }

    @Override
    public boolean onStopJob(JobParameters jobParameters) {
        return true;
    }

    private void notificationsSpotifyDF(JobParameters jobParameters) {
        compositeDisposable.add(spotifyApi.getLatestDogFood()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(disposable -> {
                    compositeDisposable.add(disposable);
                })
                .doOnComplete(() -> {
                    makeNotification(0, fullSpotifyModel);
                    jobFinished(jobParameters, true);
                })
                .subscribe(spotify -> {
                    fullSpotifyModel = new FullSpotifyModel(spotify);
                }, throwable -> {
                    Log.i(TAG, throwable.getMessage());
                }));
    }

    private void notificationsSpotifyDFC(JobParameters jobParameters) {
        compositeDisposable.add(spotifyApi.getLatestDogFoodC()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(disposable -> {
                    compositeDisposable.add(disposable);
                })
                .doOnComplete(() -> {
                    makeNotification(0, fullSpotifyModel);
                    jobFinished(jobParameters, true);
                })
                .subscribe(spotify -> {
                    fullSpotifyModel = new FullSpotifyModel(spotify);
                }, throwable -> {
                    Log.i(TAG, throwable.getMessage());
                }));
    }

    private void notificationsSpotifyOrig(JobParameters jobParameters) {
        compositeDisposable.add(spotifyApi.getLatestOrigin()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(disposable -> {
                    compositeDisposable.add(disposable);
                })
                .doOnComplete(() -> {
                    makeNotification(1, fullSpotifyModel);
                    jobFinished(jobParameters, true);
                })
                .subscribe(spotify -> {
                    fullSpotifyModel = new FullSpotifyModel(spotify);
                }, throwable -> {
                    Log.i(TAG, throwable.getMessage());
                }));
    }

    private void notificationsSpotifyOrigBeta(JobParameters jobParameters) {
        compositeDisposable.add(spotifyApi.getLatestOriginBeta()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(disposable -> {
                    compositeDisposable.add(disposable);
                })
                .doOnComplete(() -> {
                    makeNotification(1, fullSpotifyModel);
                    jobFinished(jobParameters, true);
                })
                .subscribe(spotify -> {
                    fullSpotifyModel = new FullSpotifyModel(spotify);
                }, throwable -> {
                    Log.i(TAG, throwable.getMessage());
                }));
    }

    private void makeNotification(int notificationId, FullSpotifyModel fullSpotifyModel) {
        //Launch app
        Resources resources = getResources();
        Intent i = new Intent(this, MainActivity.class)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
        PendingIntent pi = PendingIntent.getActivity(this, 0, i, 0);

        //Download spotify
        Intent intentDownload = new Intent(this, NotificationDownloadService.class);
        intentDownload.setAction(NotificationDownloadService.ACTION_DOWNLOAD);
        intentDownload.putExtra(LATEST_LINK, fullSpotifyModel.getLatestLink());
        intentDownload.putExtra(LATEST_VERSION_NAME, fullSpotifyModel.getLatestVersionName());
        intentDownload.putExtra(NOTIFICATION_ID, notificationId);
        PendingIntent piDownload = PendingIntent.getService(this, notificationId, intentDownload,
                PendingIntent.FLAG_UPDATE_CURRENT);


        //Notification
        Notification notification = new NotificationCompat.Builder(this, NOTIFICATION_CHANEL)
                .setTicker(resources.getString(R.string.app_name))
                .setSmallIcon(R.mipmap.ic_notification)
                .setContentTitle(getString(R.string.update_available))
                .setContentText(getString(R.string.new_version) + " "
                        + fullSpotifyModel.getLatestVersionName() + " " + getString(R.string.available))
                .setContentIntent(pi)
                .setAutoCancel(true)
                .setDefaults(Notification.DEFAULT_ALL)
                .setPriority(Notification.PRIORITY_HIGH)
                .setColor(getResources().getColor(R.color.colorAccent))
                .addAction(R.drawable.ic_file_download_black_24dp,
                        getString(R.string.install_now), piDownload)
                .build();

        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);

        makeNotificationChanel(notificationManager);

        showNotification(notificationId, notificationManager, notification);

    }

    private void makeNotificationChanel(NotificationManager notificationManager) {
        //Show notification on Android O
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANEL, getString(R.string.notificaion_chanel_name), NotificationManager.IMPORTANCE_HIGH);
            channel.enableLights(true);
            channel.enableVibration(true);
            channel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);
            notificationManager.createNotificationChannel(channel);
        }
    }

    private void showNotification(int notificationId, NotificationManager notificationManager, Notification notification) {
        if (notificationId == 0 && UtilsSpotify.isDogfoodUpdateAvailable(
                UtilsSpotify.getInstalledSpotifyVersion(this), fullSpotifyModel.getLatestVersionNumber())) {

            notificationManager.notify(notificationId, notification);
        } else if (notificationId == 1 && UtilsSpotify.isSpotifyUpdateAvailable(
                UtilsSpotify.getInstalledSpotifyVersion(this), fullSpotifyModel.getLatestVersionNumber())) {

            notificationManager.notify(notificationId, notification);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (compositeDisposable != null && !compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }
}

File: app/src/main/java/ru/ra66it/updaterforspotify/presentation/service/PollService.kt
package ru.ra66it.updaterforspotify.presentation.service

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.job.JobInfo
import android.app.job.JobParameters
import android.app.job.JobScheduler
import android.app.job.JobService
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.media.RingtoneManager
import android.os.Build
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import kotlinx.coroutines.*
import ru.ra66it.updaterforspotify.*
import ru.ra66it.updaterforspotify.data.storage.SharedPreferencesHelper
import ru.ra66it.updaterforspotify.domain.interactors.SpotifyInteractor
import ru.ra66it.updaterforspotify.domain.model.Result
import ru.ra66it.updaterforspotify.domain.model.SpotifyData
import ru.ra66it.updaterforspotify.presentation.ui.activity.MainActivity
import ru.ra66it.updaterforspotify.presentation.utils.SpotifyMapper
import ru.ra66it.updaterforspotify.presentation.utils.UtilsSpotify
import java.util.concurrent.TimeUnit
import javax.inject.Inject

/**
 * Created by 2Rabbit on 07.01.2018.
 */

class PollService : JobService() {
    private var job: Job? = null

    @Inject
    lateinit var spotifyInteractor: SpotifyInteractor

    @Inject
    lateinit var sharedPreferencesHelper: SharedPreferencesHelper

    @Inject
    lateinit var spotifyMapper: SpotifyMapper

    override fun onCreate() {
        super.onCreate()
        UpdaterApp.applicationComponent.inject(this)
    }

    override fun onStartJob(jobParameters: JobParameters): Boolean {
        if (sharedPreferencesHelper.isEnableNotification) {
            notificationSpotify(jobParameters)
        }
        return true
    }

    override fun onStopJob(jobParameters: JobParameters): Boolean {
        return true
    }

    private fun notificationSpotify(jobParameters: JobParameters) {
        job = CoroutineScope(Dispatchers.IO).launch {
            val result = spotifyInteractor.getSpotify()
            withContext(Dispatchers.Main) {
                when (result) {
                    is Result.Success -> {
                        val data = spotifyMapper.map(result.data)
                        makeNotification(data)
                        jobFinished(jobParameters, false)
                    }
                    is Result.Error -> {
                        if (BuildConfig.DEBUG) {
                            Log.d(TAG, result.exception.message)
                        }
                    }
                }
            }
        }
    }

    private fun makeNotification(spotifyModel: SpotifyData) {
        //Launch app
        val resources = UpdaterApp.instance.resources
        val i = Intent(this, MainActivity::class.java)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
        val pi = PendingIntent.getActivity(this, 0, i, 0)

        //Notification
        val contentText = getString(R.string.new_version) + " " +
                spotifyModel.latestVersionName + " " + getString(R.string.available)

        val havePermission = ContextCompat.checkSelfPermission(this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED

        val builder = NotificationCompat.Builder(this, notificationChanelId)
                .setTicker(resources.getString(R.string.app_name))
                .setSmallIcon(R.mipmap.ic_notification)
                .setContentTitle(getString(R.string.update_available))
                .setContentText(contentText)
                .setContentIntent(pi)
                .setAutoCancel(true)
                .setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION))
                .setDefaults(Notification.DEFAULT_ALL)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setColor(ContextCompat.getColor(this, R.color.colorAccent))

        if (havePermission) {
            //Download spotify
            val intentDownload = Intent(this, NotificationDownloadService::class.java)
            intentDownload.action = actionDownload
            intentDownload.putExtra(latestLinkKey, spotifyModel.latestLink)
            intentDownload.putExtra(latestVersionKey, spotifyModel.latestVersionNumber)
            intentDownload.putExtra(notificationIdKey, notificationId)
            val piDownload = PendingIntent.getService(this, 0, intentDownload,
                    PendingIntent.FLAG_UPDATE_CURRENT)
            builder.addAction(R.drawable.ic_file_download_black_24dp,
                    getString(R.string.install_now), piDownload)
        }

        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        makeNotificationChanel(notificationManager, builder)

        showNotification(spotifyModel.latestVersionNumber, notificationManager, builder.build())
    }

    private fun makeNotificationChanel(notificationManager: NotificationManager,
                                       builder: NotificationCompat.Builder) {
        //Show notification on Android O
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(notificationChanelId,
                    getString(R.string.notificaion_chanel_name), NotificationManager.IMPORTANCE_HIGH)
            channel.enableLights(true)
            channel.enableVibration(true)
            channel.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            builder.setChannelId(notificationChanelId)
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun showNotification(latestVersion: String, notificationManager: NotificationManager, notification: Notification) {
        if (!UtilsSpotify.isSpotifyInstalled || UtilsSpotify.isSpotifyUpdateAvailable(
                        UtilsSpotify.installedSpotifyVersion, latestVersion)) {
            notificationManager.notify(notificationId, notification)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        job?.cancel()
    }

    companion object {
        private val TAG = PollService::class.java.simpleName

        fun setServiceAlarm(isOn: Boolean, checkInterval: Long) {
            val context = UpdaterApp.instance
            val jobScheduler = context.getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
            val poolInterval = TimeUnit.DAYS.toMillis(checkInterval)

            if (isOn) {
                val needSchedule = jobScheduler.allPendingJobs.firstOrNull()?.intervalMillis != poolInterval
                if (needSchedule) {
                    val component = ComponentName(context, PollService::class.java)
                    val builder = JobInfo.Builder(jobId, component)
                            .setPeriodic(poolInterval)
                            .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
                            .setPersisted(true)
                            .build()

                    jobScheduler.schedule(builder)
                }
            } else {
                if (jobScheduler.allPendingJobs.size != 0) {
                    jobScheduler.cancelAll()
                }
            }
        }
    }
}
